### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by budiryan**
- Support nested tagging
- Change panics to errors
- Add more tests

### CODE REVIEW

**Comment by jonathanhaposan** [comment]
kyk nya ok

File: template.go [hunk 1]
```diff
@@ -274,15 +305,16 @@ func (t *Template) Execute(w io.Writer, m map[string]interface{}) (int64, error)
...
-func (t *Template) ExecuteFuncString(f TagFunc) string {
+func (t *Template) ExecuteFuncString(f TagFunc) (string, error) {
+	var s string
 	bb := t.byteBufferPool.Get()
 	if _, err := t.ExecuteFunc(bb, f); err != nil {
-		panic(fmt.Sprintf("unexpected error: %s", err))
+		return s, err
 	}
-	s := string(bb.Bytes())
+	s = string(bb.Bytes())
```
Comment (jonathanhaposan) [inline template.go]: bb typenya apa ya ?
klo buffer bytes mungkin bisa pake `.String()`
```suggestion
	s = bb.String()
```

File: template.go [hunk 2]
```diff
@@ -201,23 +205,50 @@ func (t *Template) Reset(template, startTag, endTag string) error {
 	}
 
 	for {
-		n := bytes.Index(s, a)
+		n := bytes.Index(templateBytes, startTagBytes)
```
Comment (jonathanhaposan) [inline template.go:209]: what are you trying to achive in this  loop
i guess add some guide comment in this `for {}` better

Comment (budiryan) [reply template.go:209]: ok

File: template.go (hunk 1 above)
Comment (budiryan) [reply template.go]: dia pake custom struct dia sendiri, jadi gk bisa I guess

Comment (budiryan) [reply template.go]: https://github.com/valyala/bytebufferpool

Comment (jonathanhaposan) [reply template.go]: [ini ada](https://github.com/budiryan/fasttemplate/blob/dcecefd839c4193db0d35b88ec65b4c12d360ab0/vendor/github.com/valyala/bytebufferpool/bytebuffer.go#L105)

Comment (budiryan) [reply template.go]: updated

Comment (jonathanhaposan) [reply template.go]: you forget this [one](https://github.com/budiryan/fasttemplate/pull/1/files?utf8=%E2%9C%93&diff=unified#diff-4779dc22f7a7835ae9d75daa70021ad8R97)

File: template.go (hunk 2 above)
Comment (budiryan) [reply template.go:209]: updated

File: template.go (hunk 1 above)
Comment (budiryan) [reply template.go]: updated

**Review by jonathanhaposan — APPROVED** [review]

**Pull request merged**